### PR TITLE
fix(smart-contracts): PublicLock v15 reduce contract size

### PR DESCRIPTION
**Pull request by clemsos**
# Description

A few refactoring to reduce the contract size

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 1]
```diff
@@ -209,23 +194,18 @@ contract MixinPurchase is
   ) public view returns (bool) {
     // check the lock
     if (
-      _renewalConditions[_tokenId].duration == type(uint).max ||
+      _originalDurations[_tokenId] == type(uint).max ||
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:197]:
```
_originalDurations[_tokenId].duration == type(uint).max ||
```

Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:197]:
```
_originalDurations[_tokenId].duration == type(uint).max ||
```

Comment (julien51) [reply smart-contracts/contracts/mixins/MixinPurchase.sol:197]: Please add a test so we can catch this kind of regression!

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 2]
```diff
@@ -322,7 +302,7 @@ contract MixinPurchase is
 
   function purchase(
     PurchaseArgs[] memory purchaseArgs
-  ) external payable returns (uint[] memory) {
+  ) public payable returns (uint[] memory) {
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:305]: ? Please explain :)

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 3]
```diff
@@ -404,8 +384,6 @@ contract MixinPurchase is
     address[] memory _keyManagers,
     bytes[] calldata _data
   ) external payable returns (uint[] memory) {
-    _lockPurchaseIsPossible(_recipients.length);
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:407]: Are we doing this in the Struct-purchase function?

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 4]
```diff
@@ -414,47 +392,57 @@ contract MixinPurchase is
...
-        _referrers[i],
-        _referrers[i], // here protocol referrer is the same
-        _data[i]
-      );
-      totalPriceToPay = totalPriceToPay + pricePaid;
-      tokenIds[i] = tokenId;
+      purchaseArgs[i] = PurchaseArgs({
+        value: tokenAddress != address(0) ? _values[i] : 0, //
+        recipient: _recipients[i], //
+        keyManager: _keyManagers[i], //
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:400]:
```suggestion
        value: tokenAddress != address(0) ? _values[i] : 0, 
        recipient: _recipients[i], 
        keyManager: _keyManagers[i], 
```

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 5]
```diff
@@ -414,47 +392,57 @@ contract MixinPurchase is
...
+   */
+  function _processPayment(
+    uint _tokenId,
+    address _payer,
+    address _referrer,
+    uint _pricePaid
+  ) internal {
     // receipt event
+    uint[] memory tokenIds = new uint[](1);
+    tokenIds[0] = _tokenId;
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:422]:
```suggestion
    uint[] memory tokenIds = new uint[](1);
    tokenIds[0] = _tokenId;
```

**Review by julien51 — CHANGES_REQUESTED** [review]
Please add the test!

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 6]
```diff
@@ -45,23 +45,15 @@ contract MixinPurchase is
   // default to 0
   uint256 internal _gasRefundValue;
 
-  // DEPREC: use of these is deprecated, kept for backward storage compatibility!
+  // keep track of conditions at purchase for key renewals
   mapping(uint256 => uint256) internal _originalPrices;
   mapping(uint256 => uint256) internal _originalDurations;
   mapping(uint256 => address) internal _originalTokens;
+  mapping(uint256 => address) internal _originalReferrers;
```
[the remark this answers is not in the document]
Comment (julien51) [reply smart-contracts/contracts/mixins/MixinPurchase.sol:52]: Please beware of storage!

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 7]
```diff
@@ -528,42 +492,18 @@ contract MixinPurchase is
...
-    // receipt event
-    uint[] memory tokenIds = new uint[](1);
-    tokenIds[0] = _tokenId;
-    emit PaymentReceipt(
-      tokenIds,
-      0, // purchases
-      1, // extensions
-      msg.sender, // payer
-      tokenAddress,
-      keyPrice // totalPaid
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:550]: Make sure we are using _originalPrices[_tokenId]

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 8]
```diff
@@ -528,42 +492,18 @@ contract MixinPurchase is
...
-    );
-
-    // store in unlock
-    _recordKeyPurchase(keyPrice, referrer);
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:554]: _originalPrices[_tokenId]

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 9]
```diff
@@ -528,42 +492,18 @@ contract MixinPurchase is
...
-    // refund gas if applicable
-    _refundGas();
-
-    // send what is due to referrer
-    _payReferrer(referrer);
-
-    // pay protocol
-    _payProtocol(keyPrice);
+    // pay everyone
+    _processPayment(_tokenId, ownerOf(_tokenId), referrer, keyPrice);
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol:506]:
```suggestion
    _processPayment(_tokenId, ownerOf(_tokenId), referrer, _originalPrices[_tokenId]);
```

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 3 above)
Comment (clemsos) [reply smart-contracts/contracts/mixins/MixinPurchase.sol:407]: yes

**Review by julien51 — APPROVED** [review]
ok, LGTM!

**Pull request merged**